AppActionVisualizer/app/src/main/java/com/example/appactionvisualizer/ui/adapter/ActionRecyclerViewAdapter.java

```diff
@@ -94,42 +98,54 @@ public int getItemViewType(int position) {
    */
   @Override
   public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
+    final Action action;
     switch (holder.getItemViewType()) {
-      case TYPE_ACTION:
+      case VIEW_TYPE_ACTION:
         ActionViewHolder actionHolder = (ActionViewHolder) holder;
         //search the corresponding action item index
         int idx = Arrays.binarySearch(actionPos.toArray(), position);
-        final Action action = appAction.getActions().get(idx);
+        action = actionList.get(idx);
         actionHolder.actionName.setText(action.getIntentName());
-        actionHolder.actionType.setText(action.getActionType().toString());
+        actionHolder.actionType.setText(ActionType.getActionTypeByName(action.getIntentName()).getName());
         break;
-      case TYPE_FULFILLMENT:
+      case VIEW_TYPE_FULFILLMENT:
         FulfillViewHolder fulfillHolder = (FulfillViewHolder) holder;
         //search the corresponding action item index and fulfill item index
-        int actionIdx =  (-Arrays.binarySearch(actionPos.toArray(), position)) - 2;
-        int fulfillIdx =  position - actionPos.get(actionIdx) - 1;
-        final Fulfillment fulfillment = appAction.getActions().get(actionIdx).getFulfillmentArrayList().get(fulfillIdx);
-        final String url = fulfillment.getUrlTemplate();
+        int actionIdx = (-Arrays.binarySearch(actionPos.toArray(), position)) - 2;
+        int fulfillIdx = position - actionPos.get(actionIdx) - 1;
```

Comment:
Add some comments of the logic here, e.g. why we are doing a search and transfer to negative value and minus 2?

Reply:
Comments added